Feature/sim1092

yalsayyad opened the pull request:
The core of this change is in dbConnection.py and needs a careful review.

yalsayyad committed:
Remove unused imports

yalsayyad committed:
DBObject/CatalogObject API Change
    - Initialize with driver, host, port, database
    - Remove hard-coded connection strings and addresses
    - Use dbAuth to retrieve credentials

danielsf commented on python/lsst/sims/catalogs/generation/db/dbConnection.py:
```diff
@@ -84,20 +84,45 @@ def _postprocess_results(self, chunk):
-        self.engine = create_engine(self.dbAddress, echo=self.verbose)
+
+        #DbAuth will not look up hosts that are None, '' or 0
+        if self.host:
```
Should username and password also be optional kwargs, in case users want to connect to their own databases without going through the DbAuth infrastructure?

danielsf replied:
It also seems like it might be possible for a user to write their own CatalogDBObject with the intent of connecting to their own database and then get to _connect_engine without actually specifying a host, database, etc. on account of their own forgetfulness.

Can you add a series of tests like hasattr(self,'driver') to make sure that, when _connect_to_engine() is called, the CatalogDBObject has at least enough information to make a sensible attempt at connecting?

yalsayyad replied:
I added a validation method, a more informative error message if connection fails and username/password kwargs which print a warning if you use them and throw an exception if the username is our shared username.
I'm still conflicted on the username/password kwargs: 
- We don't want to use them. 
- They are simple to add in the future, yet difficult to remove once they become used in client code.
-  Good APIs have one way to do things. 
-  Users ignore warnings.  Presence of username/password kwargs look like an invitation to use them. 
-  Passwords might creep into example scripts and demo ipython notebook repositories.

danielsf replied:
Okay.  That makes sense.  As long as the error message for a failed connection points the user to a resource (i.e. confluence) that tells them how to successfully connect to their own database using db-auth, that should be fine.

danielsf commented:
Modulo the comments above, I think this is done.

yalsayyad committed:
DBObject/Catalog API Change
    - Add username/password arguments
    - Add _validate_conn_params method to validate the database connection parameters
    - testDBObject.py: testValidationErrors()

yalsayyad committed:
Catch sqlalchemy.exc.OperationalError on connection failure
and print connection parameters

yalsayyad committed:
DBObject/Catalog API Change
    - Remove username/password arguments

yalsayyad commented:
Merged without username/password kwargs.
